src/ORM/DataObject.php

```diff
@@ -3413,7 +3448,21 @@ public function defaultSearchFilters()
      */
     public function isInDB()
     {
-        return is_numeric($this->ID) && $this->ID > 0;
+        if (empty($this->ID)) {
+            return false;
+        }
+        $class = $this->baseClass();
+        if (!isset(self::$_cache_in_db[$class][$this->ID])) {
+            $sqlSelect = new SQLSelect();
```

Comment:
The only way that you can get to this code is if you have manually set the ID and not yet written the record. So, this code is picking up the case when you've manually set the ID to an existing record.

We have the option of simply returning false in this case. If we did this, the functionality would match the "has been written" logic that I've recommended. It would remove the cause for performance / logic concerns that others have had.

My view would be that checking the database for the presence of an ID that has been manually applied is a narrow & different use-case from the current uses of `isInDB()`. In many cases where this happened it's not so much "this record is in the database" as "I am about to corrupt my database". Code responsible for ID collision would need some careful and separate thought.

So I think that should simply return false in this case, and the update the docblock of `isInDB()` to 

```
Returns true if this record was read from the database or has been written to it.

Note that it does not query the database for the presence of a manually-applied ID.
```

Note also that, after this change, the static registry can probably be replaced with an instance-specific `private boolean $isInDb` variable.